### PR TITLE
Configurable Cert Path for Auth K8s

### PR DESCRIPTION
**Pull request by JakeQuilty**
### What ticket does this PR close?
Connected to https://github.com/cyberark/conjur/issues/1559

#### Change log
- [x] The CHANGELOG has been updated, or
- [ ] This PR does not include user-facing changes and doesn't require a CHANGELOG update

#### Test coverage
- [x] This PR includes new unit and integration tests to go with the code changes, or
- [ ] The changes in this PR do not require tests

#### Documentation
- [ ] Docs (e.g. `README`s) were updated in this PR, and/or there is a follow-on issue to update docs, or
- [ ] This PR does not require updating any documentation

### CODE REVIEW

File: app/domain/authentication/authn_k8s/inject_client_cert.rb [hunk 1]
```diff
@@ -71,7 +72,7 @@ def validate
       def install_signed_cert
         pod_namespace = spiffe_id.namespace
         pod_name = spiffe_id.name
-        cert_file_path = "/etc/conjur/ssl/client.pem"
+        cert_file_path = @cert_path.nil? || @cert_path.empty? ? DEFAULT_CERT_PATH : @cert_installation_path
```
Comment (izgeri) [inline app/domain/authentication/authn_k8s/inject_client_cert.rb:75]: this logic should probably be in its own method and can be unit tested

you can see how we do this for the `host_id_prefix` (which can also be set in a header) here:
https://github.com/cyberark/conjur/blob/73c197037a3bce4122d9ae97654344e5e265e470/app/domain/authentication/authn_k8s/inject_client_cert.rb#L57-L59

maybe we can add a unit test for the logic here: https://github.com/cyberark/conjur/blob/master/spec/app/domain/authentication/authn_k8s/inject_client_cert_spec.rb?

**Review by izgeri — COMMENTED** [review]
@JakeQuilty you might find it useful to search for how the value for `host_id_prefix` is passed / parsed in a header and how its unit / integration tests are implemented - the client cert path should be similar
https://github.com/cyberark/conjur/search?q=host_id_prefix&unscoped_q=host_id_prefix

File: app/domain/authentication/authn_k8s/inject_client_cert.rb (hunk 1 above)
Comment (orenbm) [inline app/domain/authentication/authn_k8s/inject_client_cert.rb:75]: i may be missing something but where is `@cert_path` defined?

Comment (orenbm) [reply app/domain/authentication/authn_k8s/inject_client_cert.rb:75]: oh i guess it should be `@cert_installation_path`, correct?

File: config/routes.rb [hunk 2]
```diff
@@ -43,7 +43,7 @@ def matches?(request)
         # this to be accessible when using other authenticators to login.
         put  '/:authenticator/:account/api_key'  => 'credentials#rotate_api_key'
 
-        post '/authn-k8s/:service_id/inject_client_cert' => 'authenticate#k8s_inject_client_cert'
+        post '/authn-k8s/:service_id/inject_client_cert' => 'authenticate#k8s_inject_client_cert'  # <-----------------------------------
```
Comment (orenbm) [inline config/routes.rb:46]: what are all these `# <--------------`?

File: app/domain/authentication/authn_k8s/inject_client_cert.rb (hunk 1 above)
Comment (JakeQuilty) [reply app/domain/authentication/authn_k8s/inject_client_cert.rb:75]: 👍

File: config/routes.rb (hunk 2 above)
Comment (JakeQuilty) [reply config/routes.rb:46]: Yeah, I was using these just to mark important things for myself. I'm deleting these when I submit the PR as a non-WIP

File: spec/app/domain/authentication/authn_k8s/inject_client_cert_spec.rb [hunk 3]
```diff
@@ -182,6 +187,7 @@
           expect { injector.(conjur_account: account,
                             service_id: service_id,
                             csr: csr,
+                            cert_installation_path: nil_cert_installation_path,
```
Comment (jonahx) [inline spec/app/domain/authentication/authn_k8s/inject_client_cert_spec.rb:190]: I'd just get rid of `nil_cert_installation_path` and use the literal `nil`.  Seeing:
```
cert_installation_path: nil
```
will be clearer.

Comment (JakeQuilty) [reply spec/app/domain/authentication/authn_k8s/inject_client_cert_spec.rb:191]: 👍

File: app/domain/authentication/authn_k8s/inject_client_cert.rb [hunk 4]
```diff
@@ -68,10 +69,14 @@ def validate
         @validate_pod_request.(pod_request: pod_request)
       end
 
+      def set_cert_installation_path
```
Comment (izgeri) [inline app/domain/authentication/authn_k8s/inject_client_cert.rb:72]: I'll let the Ruby experts give their recommendations on this one, but my sense is that the method name shouldn't be `set_X` as that's not standard for Ruby

looking at our [style guide](https://github.com/cyberark/community/blob/master/Conjur/conventions/ruby-style-guide.md), there is indeed a link to this helpful snippet: https://github.com/rubocop-hq/ruby-style-guide#accessor_mutator_method_names

Comment (orenbm) [reply app/domain/authentication/authn_k8s/inject_client_cert.rb:72]: yes, i would change this method to:
```
def set_cert_installation_path
  @cert_installation_path ||= @cert_installation_path.empty? ? DEFAULT_CERT_PATH : @cert_installation_path
end
```

in the implementation above you don't need to explicitly check if `@cert_installation_path` is `nil` as the memoization will do it for you. please add UTs for this logic if you haven't.

Comment (orenbm) [reply app/domain/authentication/authn_k8s/inject_client_cert.rb:72]: btw, when will `request.headers["Certificate-Install-Path"]` be empty? i also don't see any UTs for this use-case?

Comment (JakeQuilty) [reply app/domain/authentication/authn_k8s/inject_client_cert.rb:72]: I was copying the same logic as https://github.com/cyberark/conjur/blob/f61ebd54b946d4210cb537049fcfcaab90a40d93/app/domain/authentication/authn_k8s/inject_client_cert.rb#L56 since they are both headers that are doing the same thing. I could see a use case as someone adding a `Certificate-Install-Path` header to the request with no value

Comment (JakeQuilty) [reply app/domain/authentication/authn_k8s/inject_client_cert.rb:72]: I'm still having trouble creating a UT for this. In the rspec tests we haven't been able to find a good way to do this. Do you have any suggestions?

File: CHANGELOG.md [hunk 5]
```diff
@@ -19,6 +19,8 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
   [cyberark/conjur#151](https://github.com/cyberark/conjur/issues/151)
 - [Documentation](./UPGRADING.md) explaining how to upgrade a Conjur server deployed in a
   Docker Compose environment. [cyberark/conjur#1528](https://github.com/cyberark/conjur/issues/1528), [cyberark/conjur#1584](https://github.com/cyberark/conjur/issues/1584)
+- Add ability to specify cert installation path on `/authn-k8s/:service_id/inject_client_cert`.
+  [cyberark/conjur#1802](https://github.com/cyberark/conjur/pull/1802)
```
Comment (izgeri) [inline CHANGELOG.md:23]: this should be linked to the issue that was filed with this request, both here and in the PR description. can you please update the changelog entry and PR description to include this link (and drop the PR link here), and also add some text to the PR description to explain what this PR does?

File: spec/app/domain/authentication/authn_k8s/inject_client_cert_spec.rb [hunk 6]
```diff
@@ -16,6 +16,9 @@
   let(:nil_host_id_prefix) { nil }
   let(:csr) { "CSR" }
 
+  let(:default_cert_file_path) { "/etc/conjur/ssl/client.pem" }
+  let(:expected_cert_file_path) { "/etc/conjur/ssl/client.pem" }
```
Comment (izgeri) [inline spec/app/domain/authentication/authn_k8s/inject_client_cert_spec.rb:20]: why is this needed? I would guess that not all tests have this as the expected cert path (you must have some tests that use a non-default entry, no?) - and when they don't, the `default_cert_file_path` can be referred to

File: spec/app/domain/authentication/authn_k8s/inject_client_cert_spec.rb [hunk 7]
```diff
@@ -371,6 +384,60 @@
...
+            allow(copy_response).to receive(:[])
+              .with(:error)
+              .and_return(expected_error_text)
+  
+            expect { injector.(conjur_account: account,
+                              service_id: service_id,
+                              csr: csr,
+                              cert_installation_path: "not_valid_path",
+                              host_id_prefix: host_id_prefix,
+                              client_ip: client_ip) }.to raise_error(error_type, expected_full_error_text)
```
Comment (izgeri) [inline spec/app/domain/authentication/authn_k8s/inject_client_cert_spec.rb:438]: these lines use different indentation than the rest of the file

**Review by izgeri — CHANGES_REQUESTED** [review]

File: app/controllers/authenticate_controller.rb [hunk 8]
```diff
@@ -120,10 +120,11 @@ def authenticator_input
...
     Authentication::AuthnK8s::InjectClientCert.new.(
-      conjur_account:   ENV['CONJUR_ACCOUNT'],
-      service_id:       params[:service_id],
-      client_ip:        request.ip,
-      csr:              request.body.read,
+      conjur_account:                ENV['CONJUR_ACCOUNT'],
+      service_id:                    params[:service_id],
+      client_ip:                     request.ip,
+      csr:                           request.body.read,
+      cert_installation_path:        request.headers["Certificate-Install-Path"],
```
Comment (orenbm) [inline app/controllers/authenticate_controller.rb:127]: should this be `Certificate-Installation-Path`?

@shulifink can you review this? it is the path in which the Conjur server will install the certificate inside the client container.

File: app/domain/authentication/authn_k8s/inject_client_cert.rb [hunk 9]
```diff
@@ -68,10 +69,14 @@ def validate
         @validate_pod_request.(pod_request: pod_request)
       end
 
+      def cert_installation_path
+        @cert_installation_path.nil? || @cert_installation_path.empty? ? DEFAULT_CERT_PATH : @cert_installation_path
+      end
```
Comment (orenbm) [inline app/domain/authentication/authn_k8s/inject_client_cert.rb:74]: Let's memoize this

```suggestion
      def cert_installation_path
        @cert_installation_path ||= @cert_installation_path.nil? || @cert_installation_path.empty? ? DEFAULT_CERT_PATH : @cert_installation_path
      end
```

Comment (orenbm) [inline app/domain/authentication/authn_k8s/inject_client_cert.rb:74]: Are you not using memoization on purpose? i think that if you'd use memoization then in case it not nil but is empty it will not work because the logic will not run. If this is the case please add a comment on why we're not using memoization so a future developer will not change it.

File: app/controllers/authenticate_controller.rb (hunk 8 above)
Comment (shulifink) [reply app/controllers/authenticate_controller.rb:127]: It's funny how each one has half abbreviated and half a full word.
Why not just use "Cert_Installation_Path" line the option?

Comment (orenbm) [reply app/controllers/authenticate_controller.rb:127]: but that's funny in the same way because now `Cert` is half abbreviated and `Installation` is a full word. I would use either `Cert-Install-Path` or `Certificate-Installation-Path`.

Regarding `cert_installation_path` - that's not customer facing but just a code variable so i wouldn't bother with its naming.

File: app/domain/authentication/authn_k8s/inject_client_cert.rb (hunk 4 above)
Comment (orenbm) [reply app/domain/authentication/authn_k8s/inject_client_cert.rb]: what if you path `cert_install_path: ""`?
